### PR TITLE
Implement Runge-Kutta 4 integrator

### PR DESCRIPTION
**Pull request by flferretti**
This PR introduces a new Runge-Kutta 4 integration method and an `Integrator` enum to select between available simulation integrators. Moreover, it adds NaN checks for quaternions and add a `pytest` fixture for testing different integrators and updates the `JaxSimModel` to support integrator selection.

----
📚 Documentation preview 📚: https://jaxsim--373.org.readthedocs.build//373/

### CODE REVIEW

File: src/jaxsim/api/integrators.py [hunk 1]
```diff
@@ -74,3 +76,77 @@ def semi_implicit_euler_integration(
...
+            joint_velocities=data_ode.joint_velocities,
+        )
+
+    def f(x) -> dict[str, jtp.Matrix]:
+
+        with data.switch_velocity_representation(jaxsim.VelRepr.Inertial):
+
+            data_ti = data.replace(
+                model=model, **{k.lstrip("_"): v for k, v in x.items()}
+            )
```
Comment (younik) [inline src/jaxsim/api/integrators.py:116]: I am afraid this will make data_ti in inertial (it never exits from the with, as we discussed previously).

If you need to put some representation in inertial, you can use dataclasses.replace on the protected attributes (that are in intertial)

File: src/jaxsim/api/integrators.py [hunk 2]
```diff
@@ -74,3 +76,77 @@ def semi_implicit_euler_integration(
...
+
+            return js.ode.system_dynamics(
+                model=model,
+                data=data_ti,
+                link_forces=link_forces,
+                joint_torques=joint_torques,
+            )
+
+    with data.switch_velocity_representation(jaxsim.VelRepr.Inertial):
+        x_t0 = get_state_derivative(data)
```
Comment (younik) [inline src/jaxsim/api/integrators.py:126]: This switch is avoidable maybe if you access the protected attributes inside the get_state_derivative function. (not big deal)

File: src/jaxsim/api/integrators.py [hunk 3]
```diff
@@ -74,3 +76,77 @@ def semi_implicit_euler_integration(
...
+
+    # Average the slopes and compute the RK4 state derivative.
+    average = lambda k1, k2, k3, k4: (k1 + 2 * k2 + 2 * k3 + k4) / 6
+
+    dxdt = jax.tree_util.tree_map(average, k1, k2, k3, k4)
+
+    # Integrate the dynamics
+    x_tf = jax.tree_util.tree_map(euler_fin, x_t0, dxdt)
+
+    data_tf = dataclasses.replace(data, **{"_" + k: v for k, v in x_tf.items()})
```
Comment (younik) [inline src/jaxsim/api/integrators.py:144]: personal opinion: I would list the attributes, as I find it more readable, even if less concise

File: src/jaxsim/api/model.py [hunk 4]
```diff
@@ -23,6 +24,13 @@
 from .common import VelRepr
 
 
+class Integrator(enum.IntEnum):
```
Comment (younik) [inline src/jaxsim/api/model.py:27]: Again, just a matter of personal opinion: "IntegratorType" sounds a better name, as Integrator would suggest to me that this is an Integrator class, i.e. implementing integrators (similar as before)

**Review by younik — COMMENTED** [review]
Good job, thanks!
I added a couple of minor comments

File: src/jaxsim/api/integrators.py (hunk 1 above)
Comment (flferretti) [reply src/jaxsim/api/integrators.py:116]: Yes, that is the intended behavior

File: src/jaxsim/api/integrators.py (hunk 2 above)
Comment (flferretti) [reply src/jaxsim/api/integrators.py:126]: Right! Thanks

Comment (flferretti) [reply src/jaxsim/api/integrators.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/373/commits/21b5ecbd3c9058faefc198f3c8534c31fe09658a

File: src/jaxsim/api/model.py (hunk 4 above)
Comment (flferretti) [reply src/jaxsim/api/model.py]: Addressed in IntegratorType

File: src/jaxsim/api/integrators.py (hunk 1 above)
Comment (younik) [reply src/jaxsim/api/integrators.py:116]: Ah okay, I see it now, we still use data later, thanks

File: src/jaxsim/api/integrators.py [hunk 5]
```diff
@@ -74,3 +76,78 @@ def semi_implicit_euler_integration(
...
+            )
+
+            return js.ode.system_dynamics(
+                model=model,
+                data=data_ti,
+                link_forces=link_forces,
+                joint_torques=joint_torques,
+            )
+
+    x_t0 = get_state_derivative(data)
```
Comment (xela-95) [inline src/jaxsim/api/integrators.py:125]: Why is the function called get_state_derivative? Doesn't that just return the input data with the normalized quaternion?

**Review by xela-95 — APPROVED** [review]
Thanks! I really like the simplicity of the implementation! Left some minor comments

File: src/jaxsim/api/ode.py [hunk 6]
```diff
@@ -158,14 +158,11 @@ def system_dynamics(
...
-    ode_state_derivative = JaxSimModelData.build(
-        model=model,
+    return dict(
         base_position=W_ṗ_B,
         base_quaternion=W_Q̇_B,
         joint_positions=ṡ,
         base_linear_velocity=W_v̇_WB[0:3],
         base_angular_velocity=W_v̇_WB[3:6],
         joint_velocities=s̈,
     )
```
Comment (xela-95) [inline src/jaxsim/api/ode.py:168]: Is this safe? Returning a dict is not ambiguous here?

Comment (flferretti) [reply src/jaxsim/api/ode.py:168]: Do you mean in the sense that the values can be changes? In that case we could use a named tuple or a similar structure. I would avoid using an additional JAX dataclass, like we used to do some months ago. The idea was to make the implementation as simple as possible.
Another option might be to make the function "private" and update the docstrings (my bad, I still need to do it)

File: src/jaxsim/api/integrators.py (hunk 3 above)
Comment (flferretti) [reply src/jaxsim/api/integrators.py]: You may find a WIP at https://github.com/ami-iit/jaxsim/pull/373/commits/79fec1290dfab08b20b993e59f63644d95064009

File: src/jaxsim/api/integrators.py (hunk 5 above)
Comment (flferretti) [reply src/jaxsim/api/integrators.py]: Removed in https://github.com/ami-iit/jaxsim/pull/373/commits/79fec1290dfab08b20b993e59f63644d95064009

File: src/jaxsim/api/ode.py (hunk 6 above)
Comment (younik) [reply src/jaxsim/api/ode.py:168]: I like the named tuple option!

**Review by younik — APPROVED** [review]

File: src/jaxsim/api/ode.py (hunk 6 above)
Comment (flferretti) [reply src/jaxsim/api/ode.py:168]: Okay, I checked better and neither NamedTuple nor MappingProxyType are valid JAX types. What I could do it to register them as valid PyTree nodes as e.g.:

```python
jax.tree_util.register_pytree_node(MappingProxyType,
    flatten_func=lambda x: list(x.items()),
    unflatten_func=lambda items: MappingProxyType(dict(items))
)
``` 

But I think it's an overkill, what about just leaving it as a dict?

Comment (younik) [reply src/jaxsim/api/ode.py:168]: Returning a dict it is okay for me

Comment (xela-95) [reply src/jaxsim/api/ode.py:168]: Yep let's proceed :)

File: src/jaxsim/api/integrators.py (hunk 3 above)
Comment (xela-95) [reply src/jaxsim/api/integrators.py]: Both option 2 and @younik one are good for me :)

Comment (flferretti) [reply src/jaxsim/api/integrators.py]: I'll leave as it is then, thank you both!
